### DISCUSSION TITLE
OctoPrint API Key management for multiple admin users

### DISCUSSION THREAD

**Discussion by lsls421** [OctoFarm/OctoFarm#510] [Implemented]
Currently, when a user API is added in Octofarm, Octofarm simply grabs the first admin user it finds from OctoPrint and tries to connect with this user account. This works in most cases smoothly without any problems, but for an Octoprint instance with more than one admin user, this could cause an error. 

Therefore, it would be a nice feature to have if the user can tell OctoFarm which admin user the API Key belongs to.

**Comment by davidzwa** [comment]
> Therefore, it would be a nice feature to have if the user can tell OctoFarm which admin user the API Key belongs to.

Not a bad idea at all as it would remove some of the confusion as well why Global API keys are not going to work: OctoFarm needs a user to which the User or Application Key belongs to be able to work with Websockets.
(WS authenticates with the string `username:key`)

On that note, @NotExpectedYet could it be that OctoPrint will fail to cooperate if we couple an API key to a username for which the key was not created? If that is the case this feature request might be of higher priority.

Repro from OctoPrint (OP) to OctoFarm (OF):
- Create 2 users: user **Ape** :monkey: and user **Banana** :banana: afterwards. 
- Make sure both **Ape** :monkey: and **Banana** :banana: are in groups _admin_ and _operator_
- Login to 2nd user **Banana** :banana:
- Create API key with **Banana** :banana: called **BananaKey** 🔑 
- Go to OF
- Add **BananaKey** 🔑 to OF as a new printer is added
- Does OP now error out or complain that **BananaKey** 🔑 does not belong to **Ape** :monkey: (which is the first user)?

**Reply by NotExpectedYet** [reply]
It does fail, the websocket will not authenticate and you don't really get much of a telling error in the UI. 

The reason I've never given this much credence before is so far I'm at 3 users who have brought this issue up. The OP been one of those been the first to actually make the request. The others said the limitation wasn't worth the effort programming and were fine with amending there setups. 

I can certainly plan the allowance for it though when I refactor the state.js file. 

I was thinking we could just alert the user there are multiple admins and they need to pick the one in the settings modal or somewhere on the UI or allow a user input in the printer add modal. If blank, auto fill as before, if filled in use that for web sockets authentication. Either of those wouldn't be too hard or time consuming to add in once that mess is cleaned up the later been what I'm more leaning too, then enabling the user field text to be updated. What do you think?

**Comment by NotExpectedYet** [comment]
> Currently, when a user API is added in Octofarm, Octofarm simply grabs the first admin user it finds from OctoPrint and tries to connect with this user account. This works in most cases smoothly without any problems, but for an Octoprint instance with more than one admin user, this could cause an error.
> 
> Therefore, it would be a nice feature to have if the user can tell OctoFarm which admin user the API Key belongs to.

Thanks for actually writing this up. No promises on a time scale but will certainly look into it more.

**Comment by NotExpectedYet** [comment]
> Currently, when a user API is added in Octofarm, Octofarm simply grabs the first admin user it finds from OctoPrint and tries to connect with this user account. This works in most cases smoothly without any problems, but for an Octoprint instance with more than one admin user, this could cause an error.
> 
> Therefore, it would be a nice feature to have if the user can tell OctoFarm which admin user the API Key belongs to.

I've allowed the user to change from all admin accounts in Octoprint in a recent commit. This will be coming in 1.2. It's currently after the printer scan but thinking of putting a text box in for printer adding going forward to submit your user before.

**Comment by NotExpectedYet** [comment]
> Currently, when a user API is added in Octofarm, Octofarm simply grabs the first admin user it finds from OctoPrint and tries to connect with this user account. This works in most cases smoothly without any problems, but for an Octoprint instance with more than one admin user, this could cause an error.
> 
> Therefore, it would be a nice feature to have if the user can tell OctoFarm which admin user the API Key belongs to.

You can now tell OctoFarm in the printer manager which user to use. You can also create an OctoFarm / octofarm user and it will automatically select that user to use. Hope this helps.
